Propegates exceptions from child to parent process.

sam-falvo opened the pull request:
This commit addresses https://github.com/cloudkick/whiskey/issues/57 .

PROBLEM

Too many times when working on Whiskey, I might make a seemingly
innocuous change to the program which causes timeouts to occur.  The
cause of the timeouts remain a total mystery.  The problem is the result
of a cascade of failures:

1) Something I changed causes an exception.
2) stdout/stderr is written to, and then the child Node process dies.
3) Meanwhile, the parent is unaware of this happening, and so sits on
its socket until it times out,
4) After the timeout, the parent process has no idea what killed the
child process, and so the only indication of a problem is that it's
listed in [TIMEOUT] status.

SOLUTION

Capture all exceptions that you can at the top-most level of the child
process, and marshal that data through the IPC channel back to the
parent process.

Then, alter the parent process to die immediately if an exception is
discovered.  Fail fast, fail often.

OTHER BENEFITS

Knowing what's causing some of the timeouts will, I _think_, be
beneficial from a Buildbot perspective as well.  No gaurantees, but it
_could_ be useful in uncovering some of the causes for intermittent
buildbot failures we've seen in the past.

HOW IT WORKS

It works by capturing all exceptions that leak to the main
run_test_file.js script, and forwards them to the _reportException
method on the current TestFile.  This stringifies the exception into a
record that is sent, via the usual Whiskey socket, to the parent Whiskey
runner process.

Once there, the lib/util.js parseResultLine method will decode this new
format and, if discovered to be an exception, will encapsulate it in a
triple, which can be uniquely identified as follows:

[true, filename, null] = File end without coverage data
[false, filenameIThink, coverageData] = File end with coverage data.
[null, testName, testResults] = Test invokation results
[string, string, null] = Exception report. (New as of this patch)

This is interpreted by the lib/run.js method
TestRunner._handleConnection(), which encapsulates the provided results
into a new exception to be handled by the parent process immediately.

The exception that is thrown is not an exact duplicate, but rather just
a simple object with a message field, formatted with text suitable for
human-readable display, as that appears to be Whiskey's default action
at this stage.  I'm not sure where the code resides that handles the
exception thrown in run.js.

Commit:
Propegates exceptions from child to parent process.

This commit addresses https://github.com/cloudkick/whiskey/issues/57 .

PROBLEM

Too many times when working on Whiskey, I might make a seemingly
innocuous change to the program which causes timeouts to occur.  The
cause of the timeouts remain a total mystery.  The problem is the result
of a cascade of failures:

1) Something I changed causes an exception.
2) stdout/stderr is written to, and then the child Node process dies.
3) Meanwhile, the parent is unaware of this happening, and so sits on
its socket until it times out,
4) After the timeout, the parent process has no idea what killed the
child process, and so the only indication of a problem is that it's
listed in [TIMEOUT] status.

SOLUTION

Capture all exceptions that you can at the top-most level of the child
process, and marshal that data through the IPC channel back to the
parent process.

Then, alter the parent process to die immediately if an exception is
discovered.  Fail fast, fail often.

OTHER BENEFITS

Knowing what's causing some of the timeouts will, I *think*, be
beneficial from a Buildbot perspective as well.  No gaurantees, but it
*could* be useful in uncovering some of the causes for intermittent
buildbot failures we've seen in the past.

HOW IT WORKS

It works by capturing all exceptions that leak to the main
run_test_file.js script, and forwards them to the _reportException
method on the current TestFile.  This stringifies the exception into a
record that is sent, via the usual Whiskey socket, to the parent Whiskey
runner process.

Once there, the lib/util.js parseResultLine method will decode this new
format and, if discovered to be an exception, will encapsulate it in a
triple, which can be uniquely identified as follows:

[true, filename, null] = File end without coverage data
[false, filenameIThink, coverageData] = File end with coverage data.
[null, testName, testResults] = Test invokation results
[string, string, null] = Exception report. (New as of this patch)

This is interpreted by the lib/run.js method
TestRunner._handleConnection(), which encapsulates the provided results
into a new exception to be handled by the parent process immediately.

The exception that is thrown is not an exact duplicate, but rather just
a simple object with a message field, formatted with text suitable for
human-readable display, as that appears to be Whiskey's default action
at this stage.  I'm not sure where the code resides that handles the
exception thrown in run.js.

jirwin commented on lib/run.js:
```diff
@@ -420,9 +420,21 @@ TestRunner.prototype._handleConnection = function(connection) {
+    var result, filePath, end, resultObj, msg, jsonPayload;
     result = testUtil.parseResultLine(line);
     end = result[0];
+    if (typeof end === 'string') {
```
I think it'd be better to use underscore here: `if (_.isString(end)) {`

sam-falvo replied:
Fixed and staged for the next commit.  It'll read underscore.isString instead, though, as that's the convention for using underscore in Whiskey.

Commit:
Use underscore instead of explicit type check

jirwin commented on lib/common.js:
```diff
@@ -296,6 +296,18 @@ TestFile.prototype.addTest = function(test) {
+    self._connection = connection;
+    callback();
+  });
+}
```
need a semicolon here for lint

jirwin commented on lib/common.js:
```diff
@@ -491,6 +489,17 @@ TestFile.prototype._getConnection = function(callback) {
+    JSON.stringify(exc),
+    constants.EXCEPTION_END_MARKER);
+  this._connection.write(payload);
+}
```
need a semicolon here for lint as well

Commit:
jshint adjustments

Kami commented on lib/run.js:
```diff
@@ -420,9 +420,21 @@ TestRunner.prototype._handleConnection = function(connection) {
+      // away.
+      jsonPayload = JSON.stringify(JSON.parse(result[1]), null, 2);
+      msg = sprintf("** EXCEPTION CAUGHT IN CHILD PROCESS **\n\n%s\n\n%s", end, jsonPayload);
+      throw {
```
If you throw something it should be an instance of `Error` class - `throw new Error(msg);`.

Throwing regular object or string is nasty and breaks a lot of things which rely on `instanceof Error`.

sam-falvo replied:
Fixed as of f62212c

Kami commented on lib/common.js:
```diff
@@ -296,6 +296,18 @@ TestFile.prototype.addTest = function(test) {
 
+TestFile.prototype.getConnection = function(callback) {
+  var self = this;
+  this._getConnection(function(err, connection) {
```
`_getConnection` always creates and establishes a new connection. The way this currently works this means we unnecessary establish two connections in the child process when we really just need one.

I would change it to do something like this
1. Add a new `_setConnection` method which uses `_getConnection` and sets `_connection` attribute on the class
2. Modify this method to call `_setConnection` if `self._connection` is not set yet. If it's already set, just proceed with the next step
3. Pass `self._connection` to the callback as the second argument

sam-falvo replied:
Fixed in c204dba

Kami commented on lib/util.js:
```diff
@@ -185,6 +187,11 @@ function parseResultLine(line) {
+  else if (exceptionIndex !== -1) {
+    exceptionString = line.split(constants.DELIMITER);
+    exceptionObj = exceptionString[1].split(constants.EXCEPTION_END_MARKER);
+    return [exceptionString[0], exceptionObj[0], null];
```
Function docstrings says `// Returns a triple (end, fileName, resultObj)`

So this branch should return like `[null, exceptionString[0], exceptionObj[0]]`.

sam-falvo replied:
That would make it indistinguishable from the other case where null is the first parameter.  Make it a 5-tuple instead?  [end, filename, result, exception-message, exception-object]?

Kami replied:
Ah, that's unfortunate. If we need to change this, we should make it better and less fragile. One of the ways to do this is to return an object from this method.

Returned object could contain the following keys:
- type - new item which represents type of the parsed data. This will allow us to distinguish between different return values.
- data - doesn't change, previously this was third item in the tuple.
- filename - doesn't change, previously this was second item in the tuple.

Example values for `type` include "test_file_end", "uncaught_exception", "coverage_object", ...

sam-falvo replied:
Fixed in f62212c

Kami commented:
This should definitely help with Whiskey development, but we need test for it.

Also a quick comment about "other benefits" section - I don't think this should in any way change or make discovering intermediate failures and timeouts easier.

Assuming that there are no bugs in Whiskey usual cause for a timeout is that some resource is not being properly closed and `test.finish()` not being called. This pull request doesn't change that.

Kami commented on lib/common.js:
```diff
@@ -491,6 +489,17 @@ TestFile.prototype._getConnection = function(callback) {
 
+TestFile.prototype._reportException = function(exc) {
+  var payload;
+  exc.message = exc.message || 'Message field not set in exception object';
```
Eventually we should re-factor code at https://github.com/sam-falvo/whiskey/blob/a3f8af9b11e8057340db2f5b5c5efca2dac9a330/lib/common.js#L171 into a separate function (e.g. `getExceptionObject` and also use it here.

sam-falvo replied:
Some fo this code is now obsolete, thanks to f62212c ; I'll be cleaning this up next.  _OOPS_  My comment applies to the wrong chunk of code.  Disregard.

sam-falvo replied:
I reviewed the the linked code -- I'm not seeing how it's relevant.  Can you explain?  Thanks.

Kami replied:
It knows how to create an `Error` object with a string message attribute even if you throw something else which is not an actual instance of `Error` class (e.g. string, etc.).

sam-falvo replied:
OH!  Duhh!  Sorry; I spaced it there.

sam-falvo replied:
Fixed in 06a227a

sam-falvo commented:
I'm not sure how to write a test for this feature without deliberately injecting a failure inside of Whiskey itself.  Alternate ideas?

Commit:
Cleanup and refactoring of parseResultLine logic.

With two sets of mutually similar if-statements in caller-side and
callee-side logic, I decided to collapse the whole shebang and use
object polymorphism to eliminate much of the control flow confusion.

The end result is:

1) Easier to expand in the future if need be, and,
2) Easier for someone fluent in OO to understand what's happening.

Commit:
Fixes built-in tests.

Commit:
Fixes getConnection() so that it runs only once.

The PR comment suggested that I alter the callback so that I pass the
connection to it.  However, doing this breaks Whiskey.  With the code as
presented, at least on my laptop, all of Whiskey's test cases pass.

Commit:
jslint error corrections

Commit:
Refactoring and reusing Error normalization code

Commit:
Refactor, code clean up, and lint fix.

sam-falvo commented:
@robert-chiniquy @Kami Now that I have this obnoxious mess sitting in github, is there any way to clean this up?  Should I just delete this branch, and start a new branch and PR?

robert-chiniquy commented:
@sam-falvo IMO keep the same PR and re-do the commitlog, you can force-push here and we have good history.

robert-chiniquy commented:
Not sure about the best way to test this.

Kami commented:
@sam-falvo @robert-chiniquy Yep, you can rebase and squash all the commits into one, although I don't really care that much. It depends on the feature scope, but personally I usually prefer writing good commit messages and just preserving the commit history (without doing a rebase) so people can see how the code evolved.

sam-falvo commented:
Just a tickle to see if this is good or not?  Thanks!

robert-chiniquy commented:
Okay so:
- @Kami I think @sam-falvo needs some guidance on how to test this
- @sam-falvo let me know, we can get together and clean up the commit history sometime soon

robert-chiniquy commented:
@Kami ^ ping
![](http://uberpong.com/wp-content/uploads/2012/10/Slow-motion-ping-pong-shot.gif)

sam-falvo commented:
Just checking for updates.  Thanks.

sam-falvo commented:
Nothing heard in five years.  Closing PR.  Someone else is free to assume control over this PR if they wish.

robertchiniquy-okta commented:
https://www.youtube.com/watch?v=ZRV32Li6h28

sam-falvo commented:
LOL!!  How are things going?  Long time no chat!

On Fri, Oct 19, 2018 at 10:31 AM Robert Chiniquy <notifications@github.com>
wrote:

> https://www.youtube.com/watch?v=ZRV32Li6h28
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/cloudkick/whiskey/pull/61#issuecomment-431438741>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABbP-Pfk6MfQWC5NDMO6CkpWkAHC05uXks5umgxqgaJpZM4AukFd>
> .
>


-- 
Samuel A. Falvo II
